### ISSUE TITLE
space-after-keywords and space-before-block give same error

### ISSUE THREAD

**Issue by nzakas** [eslint/eslint#1338]
The current implementation in #1337 has no exception for `try`, `else`, `finally`, do`.

Having `space-after-keywords` and `space-after-block` both on will raise errors like this:

```sh
34:4  error  Keyword "try" must be followed by whitespace  space-after-keywords
34:7  error  Missing space before opening brace            space-before-blocks
```

This is lousy because it means these two rules need to be configured the same, otherwise they will give opposing information. I'd like to figure out which rule should be the canonical home for this warning and remove it from the other.

**Comment by lo1tuma** [comment]
There is no `space-after-block` rule. I think you mean `space-before-blocks`.

**Comment by nzakas** [comment]
Yup, sorry for the confusion

**Comment by btmills** [comment]
Another instance of this, brought up in #3769:

```js
/* eslint "space-after-keywords": 2 */

if (a) {
    // Consequent
} else{
    // Alternate
}
```

In the demo, this will produce two errors:

```
5:7 - Keyword "else" must be followed by whitespace. (space-after-keywords)
5:7 - Missing space before opening brace. (space-before-blocks)
```

Inserting a space after the `else` will fix both messages.

In #3769, I proposed limiting space-before-blocks to cases where the block follows a closing paren because we typically have rules to cover the other cases (e.g. arrow-spacing and space-after-keywords), and this introduces duplicate or conflicting messages right now.

**Comment by lo1tuma** [comment]
> I proposed limiting space-before-blocks to cases where the block follows a closing paren

This would not work for classes:

```js
class foo {}
```

**Comment by btmills** [comment]
> This would not work for classes

Good point. I'll investigate what can precede a block and report back.

**Comment by michaelficarra** [comment]
@btmills Literally just about anything.

```js
class A extends B[C] {}
class A extends B++ {}
class A extends function(){} {}
class A extends null {}
class A extends this {}
class A extends 0 {}
// ...
```

**Comment by btmills** [comment]
Right now space-before-blocks checks three node types: `BlockStatement`, `ClassBody`, and the `case` block of a `SwitchStatement`, which doesn't have its own distinct node. As pointed out by @michaelficarra, a `ClassBody` can follow just about anything, so it doesn't make sense to apply the restriction there. What about making the restriction unique to the `BlockStatement` node type?

**Comment by nzakas** [comment]
Where would a restriction for `ClassBody` be applied?

**Comment by btmills** [comment]
Question: in a class expression, which rule should handle the space? I'd say space-after-keywords. So don't warn on `ClassBody` when the preceding token is a keyword, and only warn on `BlockStatement` when the preceding token is `(`?

**Comment by btmills** [comment]
Looks like this was unintentionally closed by a phrase in the description for #3821. The commit merged there does not reference this issue.

**Comment by nzakas** [comment]
Oops, nice catch!

**Comment by gyandeeps** [comment]
@mysticatea Can you confirm whether this issue got resolved with https://github.com/eslint/eslint/pull/4811?

**Comment by mysticatea** [comment]
I think `keyword-spacing` and `space-before-blocks` still are overlapping.
I'd like to make `space-before-blocks` ignoring spaces preceded by keywords.

**Comment by nzakas** [comment]
I think @btmills comment says the same: https://github.com/eslint/eslint/issues/1338#issuecomment-140718568

**Comment by mysticatea** [comment]
I see. I will do it on the next weekend.
